aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -82,69 +77,76 @@ Tensor empty_strided_cuda(IntArrayRef size, IntArrayRef stride, c10::optional<Sc
 Tensor& randperm_out_cuda(Tensor& result, int64_t n, c10::optional<Generator> generator) {
   TORCH_CHECK(n >= 0, "n must be non-negative, got", n);
   TORCH_CHECK(!generator.has_value() || (generator.has_value() && result.device() == generator->device()), "Expected a '", result.device(), "' generator device but found '", generator->device(), "'");
+  TORCH_CHECK(n <= std::numeric_limits<int>::max(),
+    "randperm of tensors larger than INT_MAX is not supported yet in pytorch");
   check_supported_max_int_with_precision(n, result);
 
   result.resize_({n});
 
-  if (n < 30000) {  // For small inputs, we offload it to CPU instead.
-    auto result_cpu = at::empty({n}, result.options().device(kCPU));
-    randperm_out(result_cpu, n, generator);
-    return result.copy_(result_cpu);
+  auto range = at::arange(n, result.options());
+
+  // shuffled_data points to the underlying data of the output tensor if the tensor is contiguous; otherwise it
+  // points to a new tensor.
+  Tensor shuffled;
+  void *shuffled_data;
+  if (result.is_contiguous()) {
+    shuffled_data = result.data_ptr();
+  } else {
+    shuffled = at::empty(n, result.options());
+    shuffled_data = shuffled.data_ptr();
   }
 
-#if 0
-  // This if condition should never be true because if n >= 30000 and the tensor has a Half type,
-  // check_supported_max_int_with_precision should have reported an error. This snippet is commented out but left here
-  // for the sake of clarity, because Half in thrust is spotty, and we do not want future change unaware of this.
-  if (result.scalar_type() == at::ScalarType::Half) {  // Half in thrust is spotty. Avoid.
-    auto result_float = at::empty({n}, initialTensorOptions().device(Device(DeviceType::CUDA)));
-    return result.copy_(randperm_out_cuda(result_float, n, generator));
+  auto opt = TensorOptions().device(result.device());
+
+  int bits = std::min(64,
+    static_cast<int>(std::ceil(std::log2(static_cast<double>(n)))) + 1);
+
+  if (n == 0) {
+    return result;
+  } else if (bits <= 8) {
+    auto keys = at::empty(result.sizes(), opt.dtype(kByte)).random_(generator);
```

Comment:
can you leave a comment that sorting on the first `bits` bits is still expected to produce all permutations with equal probability, provided good rng (that's true, right?)

Reply:
For infinitely mantissa random number keys, yes, but for finite mantissa, then no. For example, if your key is `(1, 0, 1)`, then your sorted value will be `(1, 0, 2)` but not `(1, 2, 0)`. So the probability of value 0 appearing in front of value 1 is always larger than the probability of value 1 appearing in front of value 0. So this algorithm has some intrinsic flaw.

Reply:
Hm, there's no mantissa, we are generating integers, no? Just random bits, and there should be all subsequences with random probability

Reply:
I will add comment tomorrow.

Reply:
But yeah, stable sort makes the argument more complex and possibly incorrect :-(

Reply:
So looks like to avoid this problem you need bits >> log2(n), to minimize probability of generating equal numbers that result in biased sorting. I wonder if this algorithm can be saved

Reply:
Yes, in theory, you need `bits == inf`. But let me theoretically analyze the probability to decide how many bits we should actually use.

Reply:
So, to sample `N` numbers from a size `M` pool, there are in total `M^N` cases. And within these cases, there are `C(M, N)` non-duplicating cases. So the probability of getting nondegenerate keys (the correct case) is
```
P = C(M,N)/(M^N)
```

Reply:
**TLDR:**
![image](https://user-images.githubusercontent.com/1032377/111667808-51314f00-87d2-11eb-9d56-f3dae9fa0081.png)


Full version: https://github.com/zasdfgbnm/things/blob/master/2021Q1/randperm.pdf

Reply:
So, roughly, if you need char n, then you need short key; if you need short n, then you need int key; if you need int n, then you need long key. If you need long n, then you are wrong, and this is not supported by cub.

Reply:
Ok, I think we can accept slightly skewed randperm :-) but you need more than `+1` that you have a as a safety margin for `bits`.

Reply:
It wouldn't be too hard to write a syncfree postprocessing kernel that additionally shuffles any islands of values whose keys match. Each thread reads its key and the keys of its neighbors to tell if it's part of an island. For each island, the "island leader" sees a key match at index i+1 but not index i-1. The island leader then reads more indices to the right (or communicates with neighboring threads through smem or shuffling) to figure out how big the island is. Most likely, the island will be very small, just a few values. The island leader then draws that many RNG and uses them to additionally shuffle values within the island and write them out.

The biggest fixed cost of this kernel would be a read of all the keys and values. Writing the reshuffled islands wouldn't be a big cost (thanks to your logic, there's a 90% chance there won't be any duplicate-key islands at all).